Add unit test cases for vfio-ioctls

jiangliu opened the pull request:
This PR prepares for publishing the vfio-ioctls by:
1) add a mock test framework
2) add unit test cases
3) support user mode device driver in addition to kvm/mshv, same goal as #33 
4) improve README.md

jiangliu committed:
ioctls: add some unit test cases

Add some unit test cases to improve code coverage.

Signed-off-by: Liu Jiang <gerry@linux.alibaba.com>

jiangliu committed:
ioctls: move VfioError into lib.rs

Move VfioError into lib.rs, and a unit test case.

Signed-off-by: Liu Jiang <gerry@linux.alibaba.com>

jiangliu committed:
ioctls: build mock framework and add test cases

Build mock framework to fake VFIO syscall and add test cases.
Support headless mode without hypervisor (kvm/mshv).

Signed-off-by: Liu Jiang <gerry@linux.alibaba.com>

jiangliu committed:
ioctls: refine README.md and Cargo.toml

Refine README.md and Cargo.toml, prepare for publishing.

Signed-off-by: Liu Jiang <gerry@linux.alibaba.com>

jiangliu committed:
ioctls: use this-error to simplify Error

Use this-error to simplify Error definition.

Signed-off-by: Liu Jiang <gerry@linux.alibaba.com>

jiangliu committed:
ioctls: introduce some helper functions

Introduce some helper functions, it will ease following patches.

Signed-off-by: Liu Jiang <gerry@linux.alibaba.com>

jiangliu committed:
ioctls: add unit test cases

Add unit test cases to improve code coverage.

Signed-off-by: Liu Jiang <gerry@linux.alibaba.com>

jiangliu committed:
ioctls: add a wrapper for VFIO ioctl syscall

Add a wrapper for VFIO ioctl syscall, so it could be overloaded
for unit tests.

Signed-off-by: Liu Jiang <gerry@linux.alibaba.com>

jiangliu committed:
ioctls: support user mode device drivers

Enhance the vfio-ioctls crate to support user mode device driver. Now
it supports three modes:
- vfio with kvm
- vfio with mshv
- vfio standalone for user mode device drivers

Signed-off-by: Liu Jiang <gerry@linux.alibaba.com>

jiangliu committed:
ioctls: fix a bug in VfioDevice::trigger_irq()

Interrupt vector should zero-based, fix it.

Signed-off-by: Liu Jiang <gerry@linux.alibaba.com>

sboeuf commented on crates/vfio-ioctls/src/lib.rs:
```diff
@@ -57,12 +57,136 @@
+                write!(f, "failed to open /dev/vfio/{} group: {}", p, e)
+            }
+            VfioError::GetGroupStatus => write!(f, "failed to get Group Status"),
+            VfioError::GroupViable => write!(f, "group is inviable"),
```
```suggestion
            VfioError::GroupViable => write!(f, "group is not viable"),
```

sboeuf commented on crates/vfio-ioctls/src/lib.rs:
```diff
@@ -57,12 +57,136 @@
+            VfioError::VfioExtension => write!(f, "failed to check VFIO extension"),
+            VfioError::VfioInvalidType => write!(f, "invalid VFIO type"),
+            VfioError::VfioType1V2 => {
+                write!(f, "container dones't support VfioType1V2 IOMMU driver type")
```
```suggestion
                write!(f, "container doesn't support VfioType1V2 IOMMU driver type")
```

sboeuf commented on crates/vfio-ioctls/src/lib.rs:
```diff
@@ -70,110 +71,55 @@ pub use vfio_device::{
+    OpenGroup(#[source] io::Error, String),
+    #[error("failed to get Group Status")]
     GetGroupStatus,
+    #[error("group is inviable")]
```
```suggestion
    #[error("group is not viable")]
```

sboeuf commented on crates/vfio-ioctls/src/lib.rs:
```diff
@@ -70,110 +71,55 @@ pub use vfio_device::{
     VfioExtension,
+    #[error("invalid VFIO type")]
     VfioInvalidType,
+    #[error("container dones't support VfioType1V2 IOMMU driver type")]
```
```suggestion
    #[error("container doesn't support VfioType1V2 IOMMU driver type")]
```

sboeuf commented on crates/vfio-ioctls/Cargo.toml:
```diff
@@ -17,6 +17,7 @@ byteorder = ">=1.2.1"
 log = "0.4"
 kvm-bindings = { version = "~0", optional = true }
 kvm-ioctls = { version = "~0", optional = true }
+thiserror = "1.0"
```
Can't we relax the requirement here with `thiserror = ">=1.0"`?

sboeuf commented on crates/vfio-ioctls/src/vfio_device.rs:
```diff
@@ -905,7 +905,8 @@ impl VfioDevice {
             .ok_or(VfioError::VfioDeviceSetIrq)?;
-        if irq.count == 0 {
+        // Currently the VFIO driver only support MASK/UNMASK INTX, so count is hard-coded to 1.
+        if irq.count == 0 || irq.count != 1 {
```
This should be in a separate commit.

sboeuf commented on crates/vfio-ioctls/src/vfio_device.rs:
```diff
@@ -791,7 +791,7 @@ impl VfioDevice {
             .ok_or(VfioError::VfioDeviceSetIrq)?;
-        if irq.count < vector {
+        if irq.count <= vector {
             return Err(VfioError::VfioDeviceSetIrq);
```
This is a general comment about `VfioError::VfioDeviceSetIrq` being used in too many places, making the error type useless because it's almost impossible to identify where the error came from.
Since you're improving the overall code with this PR, would you mind adding some new error types in order to make the error reporting better?

sboeuf commented on crates/vfio-ioctls/Cargo.toml:
```diff
@@ -14,6 +14,7 @@ mshv = ["mshv-ioctls", "mshv-bindings"]
 
 [dependencies]
 byteorder = ">=1.2.1"
+libc = "0.2.108"
```
Can't we relax this dependency?

jiangliu replied:
libc = ">=0.2.39" should work most of our cases.

sboeuf commented on crates/vfio-ioctls/src/vfio_device.rs:
```diff
@@ -906,7 +826,7 @@ impl VfioDevice {
             .ok_or(VfioError::VfioDeviceSetIrq)?;
         // Currently the VFIO driver only support MASK/UNMASK INTX, so count is hard-coded to 1.
-        if irq.count == 0 || irq.count != 1 {
+        if irq.count == 0 || irq.count != 1 || irq.index != 0 {
```
Can you please fix this in a separate commit all at once.

sboeuf commented on crates/vfio-ioctls/src/vfio_device.rs:
```diff
@@ -388,12 +390,21 @@ impl VfioGroup {
-            .write(true)
-            .open(&group_path)
-            .map_err(|e| VfioError::OpenGroup(e, id.to_string()))?;
+        let group = if cfg!(test) {
+            let tmp_file = TempFile::new().unwrap();
+            OpenOptions::new()
+                .read(true)
+                .write(true)
+                .open(tmp_file.as_path())
+                .map_err(|e| VfioError::OpenGroup(e, id.to_string()))?
+        } else {
+            let group_path = Path::new("/dev/vfio").join(id.to_string());
+            OpenOptions::new()
+                .read(true)
+                .write(true)
+                .open(&group_path)
+                .map_err(|e| VfioError::OpenGroup(e, id.to_string()))?
+        };
```
I'm not a fan of introducing test specific code in the actual codepath that's being used for non-test cases. I'd rather have two functions `new()`, one for each case, even if this duplicates a little bit more code.

sboeuf commented on README.md:
```diff
@@ -0,0 +1,27 @@
+This repository provides safe wrappers over the
+[VFIO driver framework](https://www.kernel.org/doc/Documentation/vfio.txt).
+
+Many modern system now provide DMA and interrupt remapping facilities to help ensure I/O devices behave within the boundaries they’ve been allotted. This includes x86 hardware with AMD-Vi and Intel VT-d, POWER systems with Partitionable Endpoints (PEs) and embedded PowerPC systems such as Freescale PAMU. The VFIO driver is an IOMMU/device agnostic framework for exposing direct device access to us ...
```
```suggestion
Many modern systems now provide DMA and interrupt remapping facilities to help ensure I/O devices behave within the boundaries they’ve been allotted. This includes x86 hardware with AMD-Vi and Intel VT-d, POWER systems with Partitionable Endpoints (PEs) and embedded PowerPC systems such as Freescale PAMU. The VFIO driver is an IOMMU/device agnostic framework for exposing direct device access to userspace, in a secure, IOMMU protected environment. In other words, the VFIO framework allows safe, non-privileged, userspace drivers.
```

sboeuf commented on README.md:
```diff
@@ -0,0 +1,27 @@
+
+Why do we want that? Virtual machines often make use of direct device access (“device assignment”) when configured for the highest possible I/O performance. From a device and host perspective, this simply turns the VM into a userspace driver, with the benefits of significantly reduced latency, higher bandwidth, and direct use of bare-metal device drivers.
+
+Devices are the main target of any I/O driver. Devices typically create a programming interface made up of I/O access, interrupts, and DMA. Without going into the details of each of these, DMA is by far the most critical aspect for maintaining a secure environment as allowing a device read-write access to system memory imposes the greatest risk to the overall system integrity.
```
```suggestion
Devices are the main target of any I/O driver. Devices typically create a programming interface made up of I/O accesses, interrupts, and DMA. Without going into the details of each of these, DMA is by far the most critical aspect for maintaining a secure environment as allowing a device read-write access to system memory imposes the greatest risk to the overall system integrity.
```

sboeuf commented on README.md:
```diff
@@ -0,0 +1,27 @@
+
+To help mitigate this risk, many modern IOMMUs now incorporate isolation properties into what was, in many cases, an interface only meant for translation (ie. solving the addressing problems of devices with limited address spaces). With this, devices can now be isolated from each other and from arbitrary memory access, thus allowing things like secure direct assignment of devices into virtual mac ...
+
+While for the most part an IOMMU may have device level granularity, any system is susceptible to reduced granularity. The IOMMU API therefore supports a notion of IOMMU groups. A group is a set of devices which is isolatable from all other devices in the system. Groups are therefore the unit of ownership used by VFIO.
```
```suggestion
While for the most part an IOMMU may have device level granularity, any system is susceptible to expose a reduced granularity. The IOMMU API therefore supports a notion of IOMMU groups. A group is a set of devices which is isolated from all other devices in the system. Groups are therefore the unit of ownership used by VFIO.
```

sboeuf commented on README.md:
```diff
@@ -0,0 +1,27 @@
+
+## Usage
+This repository provides two crates to use the VFIO framework, please refer to crate documentations for detail information.
+- [vfio-bindings](https://github.com/jiangliu/vfio-ioctls/tree/ioctls/crates/vfio-bindings): a rust FFI bindings to VFIO generated using [bindgen](https://crates.io/crates/bindgen).
```
```suggestion
- [vfio-bindings](https://github.com/rust-vmm/vfio-ioctls/tree/ioctls/crates/vfio-bindings): a rust FFI bindings to VFIO generated using [bindgen](https://crates.io/crates/bindgen).
```

sboeuf commented on README.md:
```diff
@@ -0,0 +1,27 @@
+## Usage
+This repository provides two crates to use the VFIO framework, please refer to crate documentations for detail information.
+- [vfio-bindings](https://github.com/jiangliu/vfio-ioctls/tree/ioctls/crates/vfio-bindings): a rust FFI bindings to VFIO generated using [bindgen](https://crates.io/crates/bindgen).
+- [vfio-ioctls](https://github.com/jiangliu/vfio-ioctls/tree/ioctls/crates/vfio-ioctls): a group of safe wrappers over the [VFIO APIs](https://github.com/torvalds/linux/blob/master/include/uapi/linux/vfio.h).
```
```suggestion
- [vfio-ioctls](https://github.com/rust-vmm/vfio-ioctls/tree/ioctls/crates/vfio-ioctls): a group of safe wrappers over the [VFIO APIs](https://github.com/torvalds/linux/blob/master/include/uapi/linux/vfio.h).
```

sboeuf commented on crates/vfio-ioctls/README.md:
```diff
@@ -1,41 +1,68 @@
-- What are the main components of the crate? How do they interact which each
-  other?
+The VFIO framework is originally developed on Linux system, and later Microsoft HyperV
+technology provides an compatible implementation. So the VFIO framework is supported
```
```suggestion
technology provides a compatible implementation. Therefore the VFIO framework is supported
```

sboeuf reviewed (commented):
Very nice PR :+1: 
I've tested it with Cloud Hypervisor and I confirm it works correctly :)

jiangliu committed:
ioctls: only support mask/unmask for legacy irq


Currently the VFIO driver only support MASK/UNMASK INTX, so limit
mask()/unmask() to legacy IRQ only.

Signed-off-by: Liu Jiang <gerry@linux.alibaba.com>

jiangliu committed:
ioctls: improve error codes for IRQ operations

Currently all IRQ operations share a single error code, it's inconvient
for debugging. So introduce dedicated error codes for each IRQ
operation.

Signed-off-by: Liu Jiang <gerry@linux.alibaba.com>

sboeuf reviewed (commented):
@jiangliu 
Thanks for the quick update of your PR.
I don't think you addressed the comments on `libc` and `thiserror` dependencies. Just wanted to have your thoughts on these.

jiangliu commented:
> @jiangliu Thanks for the quick update of your PR. I don't think you addressed the comments on `libc` and `thiserror` dependencies. Just wanted to have your thoughts on these.

It has been updated by the last patch, I will move the modification to the right patch.

andreeaflorescu commented on crates/vfio-ioctls/src/fam.rs:
```diff
@@ -33,7 +40,50 @@ fn vec_with_size_in_bytes<T: Default>(size_in_bytes: usize) -> Vec<T> {
 pub(crate) fn vec_with_array_field<T: Default, F>(count: usize) -> Vec<T> {
-    let element_space = count * size_of::<F>();
-    let vec_size_bytes = size_of::<T>() + element_space;
+    let element_space = match count.checked_mul(size_of::<F>()) {
```
Why is this still needed? I noticed you're using the fam struct from vmm-sys-util as well.

jiangliu replied:
It may be replaced by FamStruct, I hope do that with a separate PR:)

jiangliu replied:
I have double checked the code, and it seems FamStruct doesn't help much here. The structure used with `vec_with_array_field` have conflicts with FamStruct's assumption.
1) those structs has field "argsz" (in bytes) instead of `n_elements`
2) the following elements may be different type/size.

Need more thinking about whether it worth to switching to FamStruct.

liuw commented on crates/vfio-ioctls/README.md:
```diff
@@ -1,41 +1,68 @@
-- What is the purpose of this crate?
-- What are the main components of the crate? How do they interact which each
-  other?
+The VFIO framework is originally developed on Linux system, and later Microsoft HyperV
```
`Microsoft HyperV` should be `Microsoft Hypervisor`

liuw commented on crates/vfio-ioctls/README.md:
```diff
@@ -1,41 +1,68 @@
-  other?
+The VFIO framework is originally developed on Linux system, and later Microsoft HyperV
+technology provides a compatible implementation. Therefore the VFIO framework is supported
+by both Linux and Windows HyperV.
```
Same here.

liuw commented on crates/vfio-ioctls/README.md:
```diff
@@ -1,41 +1,68 @@
-TODO: This section describes how the crate is used.
+The `vfio-ioctls` crate may be used to support following usage scenarios:
+- Direct device assignment to virtual machine based on Linux KVM, with default features.
+- Direct device assignment to virtual machine based on Microsoft HyperV, with `--no-default-features --features=mshv`.
```
Same here.

liuw commented on crates/vfio-ioctls/src/vfio_device.rs:
```diff
@@ -309,13 +254,124 @@ impl VfioContainer {
+}
+
+#[cfg(all(feature = "mshv", not(feature = "kvm"), not(test)))]
+// Methods to support the Microsoft HyperV hypervisor.
```
Just use `Microsoft Hypervisor` here.

sboeuf commented:
@jiangliu now this is merged, are you planning on updating the repository name to `vfio`?

liuw commented:
> @jiangliu now this is merged, are you planning on updating the repository name to `vfio`?

Can we make a new copy? Making the old one disappear will break builds and make code archaeology much harder.

jiangliu commented:
> @jiangliu now this is merged, are you planning on updating the repository name to `vfio`?

@andreeaflorescu any suggestions here? How about creating a new vfio repo and archives the vfio-bindings/vfio-ioctls repos? We could clone from vfio-ioctls into vfio repo to keep all history info.

andreeaflorescu commented:
I created https://github.com/rust-vmm/vfio. You can set it up as discussed here.
